### ISSUE TITLE
Improve usability and extensibility of Link Editor

### ISSUE THREAD

**Issue by bwaidelich** [neos/neos-ui#2669]
### Description

The Link Editor is one of the most crucial editor in a CMS.
Ours is currently very straightforward but IMO it lacks some important features.

#### Missing features

* It's not easy to find the correct target page for a site with many pages
* Cross-Site links are not transparent/possible
* It's not possible to set a link's title text
* It's not possible to set the link target (`_blank` for example)
* It's not possible to create anchor links
* It's not (easily) possible to create email links
* ..and probably some more

In general there are some UX issues with the current implementation in my opinion:

##### Creating Links 1

![image](https://user-images.githubusercontent.com/307571/74144124-71b8ce00-4bfc-11ea-9d13-3470687f5f83.png)

* The node URL underneath each item in the result is pretty useless (even without the "/neos/preview" part it was usually too long to be helpful)
* By default assets just appear in the results which is rather confusing and not so easy to extend upon (see below)

##### Creating Links 2

![image](https://user-images.githubusercontent.com/307571/74144295-c2c8c200-4bfc-11ea-926d-e06adc10e749.png)

* There is no (easy) way to know what node is linked – especially in large sites there might be duplicates in the title.
* There is no way to jump to the target page from here
* The preview URL is not helpful here either

##### Editing Links

![image](https://user-images.githubusercontent.com/307571/74144451-07ecf400-4bfd-11ea-9eae-efbc72e5ba9c.png)

"Dear editor, this is how we internally store page links. Deal with it" :)

#### No easy extensibility

Lastly, it's currently not a trivial task at all to extend the current Link Editor.
It has been [done](https://github.com/di-unternehmer/DIU.Neos.AnchorLink) by @dimaip for example  but there is no API.

**Comment by bwaidelich** [comment]
I would suggest to make "link types" an explicit (and extensible) concept.
This would allow us to provide (or reuse) a dedicated editor for each type:

![image](https://user-images.githubusercontent.com/307571/74145020-2ef7f580-4bfe-11ea-938d-3d3283da3382.png)

![image](https://user-images.githubusercontent.com/307571/74145053-37503080-4bfe-11ea-9021-d370246ea425.png)

Also folks could extend/limit the allowed link types to be used, for example to provide an easy way to link to product pages

**Comment by Sebobo** [comment]
Your scribble is basically the good old TYPO3 linkeditor ;)

![Bildschirmfoto 2020-02-10 um 12 31 11](https://user-images.githubusercontent.com/596967/74146383-45538080-4c01-11ea-929b-26e64a5f8124.png)

But I agree, it would help make things clearer. 

I think the biggest gain would be to "translate" the linkurl and suggestions to something human readable and give more space. Because that would also help in the inspector etc.

I'm not super into the tabs somehow.

**Comment by bwaidelich** [comment]
> Your scribble is basically the good old TYPO3 linkeditor ;)

Yep, that's what I had in the back of my head :)
Though the TYPO3 link editor is a little too low-levely IMO (CSS-Class, Link Parameters) and the "link type" API inofficial IIRC.

> I'm not super into the tabs somehow

Yep, that's why I just created a mockup. I don't even want to _try_ to come up with a good UI for this. But it should be a different editor per link type IMO

**Comment by lorenzulrich** [comment]
Maybe you meant this Discuss post: https://discuss.neos.io/t/initiative-improve-search-results-in-reference-editor/4622/2

**Comment by bwaidelich** [comment]
> Maybe you meant this Discuss post

Yes, I think that was it. Thanks for linking!

**Comment by mficzel** [comment]
I like the concept but not the tabs. They also could lead to conceptually errors as tabs lie in the inspector usually allow to fill values on multiple tabs which is no valid usecase.

It makes sense to specify what to link first (with default to nodes) and then show optimized controls. That would probably mean a select box or radios buttons for the protocol (http://, node://, mailto://, resource:). 

On the long run the protocol list could be made extensible but for that the linkingService would have to be extended first.

**Comment by bwaidelich** [comment]
> I like the concept but not the tabs

I agree. In the mockups above the tabs should just represent multiple exclusive contexts

>  It makes sense to specify what to link first

Yes, that's true. Tabs didn't really communicate that. It's probably more like a 2-step wizard (if more than one link type is allowed)

**Comment by bweinzierl** [comment]
I think the most pressing issue is the inability to link nodes form another site (without using absolute links). That is a big issue for every site that uses neos in a multi domain context. Basically this is currently not really possible without cluttering the database with static links.
It can be configured for all inspector editors like this, but not for the CKEditor Link Editor:
```
Neos:
  Neos:
    userInterface:
      inspector:
        editors:
          Neos.Neos/Inspector/Editors/ReferenceEditor:
            editorOptions:
              startingPoint: '/sites'
```

And a big +1 for the extensible concept of link types. It would be great if we could keep the configuration options relatively in sync (as far as possible) with the options of the Reference Editor

**Comment by Benjamin-K** [comment]
> It can be configured for all inspector editors like this, but not for the CKEditor Link Editor

@bweinzierl: Actually I think that this is not the case. You can set the startingPoint for links in the CKEditor in the NodeType definition like this:

```yaml
'Vendor.Package:NodeType':
  properties:
    text:
      ui:
        inline:
          editorOptions:
            linking:
              startingPoint: '/sites'
```

However, it is not (yet) possible to use ClientEval there, I already opened an issue for that.

**Comment by bwaidelich** [comment]
FYI: https://github.com/sitegeist/Sitegeist.Archaeopteryx implements a great deal of features in a separate package

**Comment by lorenzulrich** [comment]
Wow, that's absolutely great!

**Comment by creative-resort** [comment]
Is Archaeopteryx a candidate for v8.0?

@grebaldi

**Comment by iseries** [comment]
Archaeopteryx looks really cool!

**Comment by mficzel** [comment]
If someone is willing to do the work of moving  Archeaopteryx we @sitegeist would be fine with that. Was always the plan in the long run. There are some changes that we would recommend still. It would require solid fe-skills, I can only support but am no fe-dev.

**Comment by markusguenther** [comment]
Hmm, interesting. Can not promise but will look into it.

**Comment by mhsdesign** [comment]
https://discuss.neos.io/t/rfc-move-sitegeist-archaeopteryx-link-editor-to-the-core/6024

**Comment by mhsdesign** [comment]
the Archaeopteryx link editor will now be part of 9.1 via https://github.com/neos/neos-ui/pull/3996
it also got a little visual overhaul;)

I think we can close this then ... feel free to test it already!
